### ISSUE TITLE
GetCustomAttributes on a type from an assembly in the LoadFile context fails

### ISSUE THREAD

**Issue by rjvdboon** [mono/mono#8575]
## Steps to Reproduce

1. Extract the attached repro
2. `make run`

[Repro.zip](https://github.com/mono/mono/files/1974822/Repro.zip)

### Current Behavior

It throws an Exception.

### Expected Behavior

Console shows `Found 1 MyCustomAttribute attributes`

## On which platforms did you notice this

[ ] macOS
[X ] Linux
[ ] Windows

The repro case gives the expected output on .Net framework and Mono up to 5.12.0.128.
It gives the exception on mono/master.

**Version Used**:

```
$ mono --version
Mono JIT compiler version 5.15.0 (master/dbfbffb559a Fri May  4 13:57:34 CEST 2018)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       altstack
        Notifications: epoll
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug
        Interpreter:   yes
        LLVM:          supported, not enabled.
        GC:            sgen (concurrent by default)
$ gcc --version | head -1
gcc (GCC) 4.4.7 20120313 (Red Hat 4.4.7-17)
$ uname -srvm
Linux 2.6.32-696.3.2.el6.x86_64 #1 SMP Wed Jun 7 11:51:39 EDT 2017 x86_64
```

## Stacktrace

```
Can't find custom attr constructor image: /scratch/repro/A.dll mtoken: 0x0a000001 due to: Could not load file or assembly 'Attr, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies.
Got an exception: System.IO.FileNotFoundException: Could not load file or assembly 'Attr, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies.
File name: 'Attr, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'
  at (wrapper managed-to-native) System.MonoCustomAttrs.GetCustomAttributesDataInternal(System.Reflection.ICustomAttributeProvider)
  at System.MonoCustomAttrs.GetCustomAttributesData (System.Reflection.ICustomAttributeProvider obj) [0x0000e] in /scratch/mono/mono-git/mcs/class/corlib/System/MonoCustomAttrs.cs:303
  at System.Reflection.CustomAttributeData.GetCustomAttributesInternal (System.RuntimeType target) [0x00000] in /scratch/mono/mono-git/mcs/class/corlib/System.Reflection/CustomAttributeData.cs:123
  at System.RuntimeType.GetCustomAttributesData () [0x00000] in /scratch/mono/mono-git/mcs/class/referencesource/mscorlib/system/rttype.cs:5119
  at System.Reflection.MemberInfo.get_CustomAttributes () [0x00000] in /scratch/mono/mono-git/mcs/class/referencesource/mscorlib/system/reflection/memberinfo.cs:51
  at P.Prog.Main () [0x0006e] in <f898189524a448b8a9de610606c5ddab>:0
```

**Comment by lambdageek** [comment]
Somehow Attr.dll gets loaded in the default context in .NET Framework.

(It's not that A.dll is loaded into the default context - other referenced assemblies (for example adding a class with a base type in a referenced assembly) are not loaded even when needed)

**Comment by MaximLipnin** [comment]
The similar behavior can be be reproduced on the .Net framework if there is no Attr.dll in bin folder.
On Mono master the repro gives expected result after replacing Assembly.LoadFile with Assembly.LoadFrom.

**Comment by lambdageek** [comment]
This will be fixed by #9858.  assemblies loaded by ~LoadFile~ are not as isolated as I thought - they still can reference assemblies from the GAC, and the application base and private bin path of the appdomain.
